{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Assignment1 -Ch2(2016331092).ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "H71heObrRXlO"
      },
      "source": [
        "#### BA2A: Implement MotifEnumeration"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lmmzKgfyQLaW",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "909d6222-c829-4bd4-b0ec-7f06239d147d"
      },
      "source": [
        "def generateAllKmers(k):\n",
        "    base_neucleotides = [\"A\", \"T\", \"G\", \"C\"]\n",
        "    last_base = base_neucleotides #copying base neucleotides in last_base\n",
        "    current_array = [] #declare an emty list\n",
        "    for i in range(k-1): \n",
        "        for j in base_neucleotides: #A, T, G, C\n",
        "            for k in last_base: #copy of A, T, G, C\n",
        "                current_array.append(j+k) \n",
        "        last_base = current_array #update last_base\n",
        "        current_array = [] #empty current array\n",
        "    return last_base\n",
        "\n",
        "def MismatchedMotifSet(text, k, d):\n",
        "    motifset = set()\n",
        "    all_Kmer = generateAllKmers(k) #generate all possible k mers\n",
        "    text_Kmer = textKmer(text, k)  #generate all possible kmers from given text\n",
        "    for i in range(len(all_Kmer)):\n",
        "        for j in range(len(text_Kmer)):\n",
        "          if hammingDistance(all_Kmer[i], text_Kmer[j]) <= d: #for each kmer generated from all possible kmers check hamming distance with main text kmers\n",
        "            motifset.add(all_Kmer[i]) \n",
        "    return motifset\n",
        "    \n",
        "def hammingDistance(text1, text2):\n",
        "    count=0;\n",
        "    for i in range(len(text1)):\n",
        "        if(text1[i] != text2[i]):\n",
        "            count+=1\n",
        "    return count\n",
        "\n",
        "def textKmer(text, k):\n",
        "    my_arr = []\n",
        "    for i in range(len(text)-k+1):\n",
        "        my_arr.append(text[i:i+k])\n",
        "    return my_arr\n",
        "\n",
        "def DNeighbourhood(genomelist, k,d):\n",
        "  count = {}\n",
        "\n",
        "  for genome in genomelist:\n",
        "    motifset = MismatchedMotifSet(genome, k, d)\n",
        "    for motif in motifset:\n",
        "      if motif not in count:\n",
        "        count[motif] = 1\n",
        "      else:\n",
        "        count[motif] += 1\n",
        "  \n",
        "  motiflist = []\n",
        "  for key in count:\n",
        "    if count[key] == len(genomelist):\n",
        "      motiflist.append(key)\n",
        "  return sorted(motiflist)\n",
        "\n",
        "def WriteToFile(result):\n",
        "  outF = open(\"myOutFile.txt\", \"w\")\n",
        "  for line in result:\n",
        "    # write line to output file\n",
        "    outF.write(line)\n",
        "    outF.write(\" \")\n",
        "  outF.close()\n",
        "\n",
        "k = 3\n",
        "d = 1\n",
        "with open('/content/rosalind_ba2a.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "length = len(genomelist)\n",
        "ans = DNeighbourhood(genomelist, k, d)\n",
        "print(ans)\n"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "['ATA', 'ATT', 'GTT', 'TTT']\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WEj-LvMPR7il"
      },
      "source": [
        "#### BA2B: Find a Median String"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MLBIRhtmR7rr",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ed3ba6aa-700a-45b0-a291-510d777b9237"
      },
      "source": [
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): #base case\n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem] #remainder is the single symbol\n",
        "    pref_pattern = NumberToPattern(pref_index, k-1) #recursive call\n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "def HammingDistance(text1, text2):\n",
        "    count=0;\n",
        "    for i in range(len(text1)):\n",
        "        if(text1[i] != text2[i]):\n",
        "            count+=1\n",
        "    return count\n",
        "\n",
        "def DistanceBetweenPatternsandGenome(genomelist, pattern):\n",
        "  k = len(pattern)\n",
        "  distance = 0\n",
        "  for genome in genomelist:\n",
        "    hammingDistance = 9999\n",
        "    genome_kmers = generateTextKmer(genome, k)\n",
        "    for kmer in  genome_kmers:\n",
        "      if hammingDistance > HammingDistance(pattern, kmer):\n",
        "        hammingDistance =  HammingDistance(pattern, kmer)\n",
        "    distance = distance+hammingDistance\n",
        "  #print(distance)  \n",
        "  return distance\n",
        "\n",
        "def MedianString(genomelist, k):\n",
        "  distance = 9999;\n",
        "  limit = pow(4,k)\n",
        "  for i in range(limit-1):\n",
        "    pattern = NumberToPattern(i,k)\n",
        "    if distance >= DistanceBetweenPatternsandGenome(genomelist, pattern):\n",
        "      #print(\"pattern, Distance\")\n",
        "      #print(pattern, distance)\n",
        "      distance = DistanceBetweenPatternsandGenome(genomelist, pattern)\n",
        "      median = pattern\n",
        "  return median\n",
        "\n",
        "k = int(input(\"Enter the length of kmer: \"))\n",
        "genomelist = []\n",
        "while True:\n",
        "  line = input().upper()\n",
        "  if line:\n",
        "    genomelist.append(line)\n",
        "  else:\n",
        "    break\n",
        "#print(genomelist)\n",
        "ans = MedianString(genomelist, k)\n",
        "print(ans)\n",
        "\n"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Enter the length of kmer: 3\n",
            "AAATTGACGCAT\n",
            "GACGACCACGTT\n",
            "CGTCAGCGCCTG\n",
            "GCTGAGCACCGG\n",
            "AGTACGGGACAG\n",
            "\n",
            "GAC\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Bkp10jJWR73S"
      },
      "source": [
        "#### BA2C: Find a Profile-most Probable k-mer in a String"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "AhgMF5cdR7_b",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 103
        },
        "outputId": "dd57cbb0-1865-4302-ae15-8a3550984a70"
      },
      "source": [
        "import numpy as np\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "def ProfileMostProbableKmer(genome, k, profile_matrix):\n",
        "  kmers = generateTextKmer(genome, k)\n",
        "  maxProb = -9999\n",
        "  probKmer = \"\"\n",
        "  for kmer in kmers:\n",
        "    prob = 1\n",
        "    for i in range(len(kmer)):\n",
        "      if kmer[i] == 'A':\n",
        "        prob = prob*profile_matrix[0][i]\n",
        "      elif kmer[i] == 'C':\n",
        "        prob = prob*profile_matrix[1][i]\n",
        "      elif kmer[i] == 'G':\n",
        "        prob = prob*profile_matrix[2][i]\n",
        "      elif kmer[i] == 'T':\n",
        "        prob = prob*profile_matrix[3][i]\n",
        "    if maxProb < prob:\n",
        "      maxProb = prob\n",
        "      probKmer = kmer\n",
        "\n",
        "  return probKmer\n",
        "\n",
        "\n",
        "genome = input(\"Enter the genome: \")\n",
        "k = int(input(\"Enter pattern length: \"))\n",
        "print(\"Enter the entries in a single line separated by space: \")\n",
        "arr_entry1 = list(map(float, input().strip().split()))\n",
        "profile_matrix =  np.array(arr_entry1).reshape(4,k)\n",
        "ProfileMostProbableKmer(genome, k, profile_matrix)"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Enter the genome: ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT\n",
            "Enter pattern length: 5\n",
            "Enter the entries in a single line separated by space: \n",
            "0.2 0.2 0.3 0.2 0.3 0.4 0.3 0.1 0.5 0.1 0.3 0.3 0.5 0.2 0.4 0.1 0.2 0.1 0.1 0.2\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'CCGAG'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Cpoi50_wR8F9"
      },
      "source": [
        "#### BA2D: Implement GreedyMotifSearch"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Bk5u0oSuR8MG",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "902454d8-5d6e-4538-86ed-70988edb9f27"
      },
      "source": [
        "import numpy as np\n",
        "\n",
        "def BestMotifs(genomes, k,t):\n",
        "  bestmotifs = []\n",
        "  for genome in genomes:\n",
        "    bestmotifs.append(genome[:k])\n",
        "  return bestmotifs\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "def CountMatrix(motifs): \n",
        "  count = np.zeros((4,len(motifs[0])))\n",
        "  for motif in motifs:\n",
        "    for i in range(len(motif)):\n",
        "      if motif[i-1] == 'A':\n",
        "        count[0][i-1] += 1\n",
        "      elif motif[i-1] == 'C':\n",
        "        count[1][i-1] += 1\n",
        "      elif motif[i-1] == 'G':\n",
        "        count[2][i-1] += 1\n",
        "      elif motif[i-1] == 'T':\n",
        "        count[3][i-1] += 1\n",
        "  return count\n",
        "\n",
        "def ProfileMatrix(motifs): \n",
        "  countMatrix = CountMatrix(motifs)\n",
        "  profileMatrix = np.zeros((4,len(motifs[0])))\n",
        "  #divide count matrix by number of motifs\n",
        "  for row in range(4):\n",
        "    for col in range(k):\n",
        "      profileMatrix[row][col] = countMatrix[row][col] / len(motifs)\n",
        "  return profileMatrix\n",
        "\n",
        "\n",
        "def ProfileMostProbableKmer(genome, k, profile_matrix):\n",
        "  maxprob = 0\n",
        "  kmer = genome[0:k]\n",
        "  for i in range(1,len(genome) - k +1):\n",
        "    prob = 1\n",
        "    pattern = genome[i:i+k]\n",
        "    for j in range(k):\n",
        "      l = symbolToNumber(pattern[j])\n",
        "      prob *= profile_matrix[l][j]\n",
        "    if prob > maxprob:\n",
        "      maxprob = prob\n",
        "      kmer = pattern\n",
        "  return kmer\n",
        "  \n",
        "def symbolToNumber(symbol):\n",
        "\tif symbol == \"A\":\n",
        "\t\treturn 0\n",
        "\tif symbol == \"C\":\n",
        "\t\treturn 1\n",
        "\tif symbol == \"G\":\n",
        "\t\treturn 2\n",
        "\tif symbol == \"T\":\n",
        "\t\treturn 3\n",
        "\n",
        "def NumberToSymbol(number):\n",
        "\tif number == 0:\n",
        "\t\treturn \"A\"\n",
        "\tif number == 1:\n",
        "\t\treturn \"C\"\n",
        "\tif number == 2:\n",
        "\t\treturn \"G\"\n",
        "\tif number == 3:\n",
        "\t\treturn \"T\"\n",
        "\n",
        "def Consensus(profile):\n",
        "  consensus = \"\"\n",
        "  for col in range(len(profile[0])):  \n",
        "    max = 0\n",
        "    loc = 0\n",
        "    for row in range(4):\n",
        "      if profile[row][col] > max:\n",
        "        loc = row  \n",
        "        max = profile[row][col]  \n",
        "    consensus += NumberToSymbol(loc)  \n",
        "  return consensus\n",
        "\n",
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): \n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem] \n",
        "    pref_pattern = numberToPattern(pref_index, k-1) \n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def Score(motifs):\n",
        "  profile = ProfileMatrix(motifs)\n",
        "  cons = Consensus(profile)\n",
        "  score = 0\n",
        "  for motif in motifs:  \n",
        "    for i in range(len(motif)): \n",
        "      if cons[i] != motif[i]:\n",
        "        score+=1\n",
        "  return score\n",
        "\n",
        "def GreedyMotif(genomes, k, t):\n",
        "  bestMotifs = BestMotifs(genomes, k,t)\n",
        "  kmers = generateTextKmer(genomes[0], k)\n",
        "  for kmer in kmers:\n",
        "    motifs = []\n",
        "    motifs.append(kmer)\n",
        "    for idx in range(1, t):\n",
        "      profile_matrix = ProfileMatrix(motifs) \n",
        "      motifs.append(ProfileMostProbableKmer(genomes[idx], k, profile_matrix))\n",
        "    if Score(motifs) < Score(bestMotifs):\n",
        "      bestMotifs = motifs\n",
        "  return bestMotifs\n",
        "\n",
        "k = 3\n",
        "t = 5\n",
        "with open('/content/rosalind_ba2d.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "ans = GreedyMotif(genomelist, k, t)\n",
        "for idx in ans:\n",
        "  print(idx)"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "CAG\n",
            "CAG\n",
            "CAA\n",
            "CAA\n",
            "CAA\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_n-RAldzR8Ti"
      },
      "source": [
        "#### BA2E: \n",
        "Implement GreedyMotifSearch with Pseudocounts"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TplSRZtkR8b7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "58f98147-80aa-44fd-a9a8-987d8bfc795f"
      },
      "source": [
        "import numpy as np\n",
        "def BestMotifs(genomes, k,t):\n",
        "  bmotifs = []\n",
        "  for genome in genomes:\n",
        "    bmotifs.append(genome[:k])\n",
        "  return bmotifs\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "\n",
        "def CountMatrix(motifs): # count matrix is 4*k AAGG\n",
        "  count = np.zeros((4,len(motifs[0])))\n",
        "  for motif in motifs:\n",
        "    #print(motif)\n",
        "    for i in range(len(motif)):\n",
        "      # print(\"i: \", i)\n",
        "      # print(motif[i])\n",
        "      if motif[i-1] == 'A':\n",
        "        count[0][i-1] += 1\n",
        "      elif motif[i-1] == 'C':\n",
        "        count[1][i-1] += 1\n",
        "      elif motif[i-1] == 'G':\n",
        "        count[2][i-1] += 1\n",
        "      elif motif[i-1] == 'T':\n",
        "        count[3][i-1] += 1\n",
        "  return count\n",
        "\n",
        "def ProfileMatrix(motifs): \n",
        "  countMat = CountMatrix(motifs)\n",
        "  pseudoMat = PseudoCountMatrix(countMat)\n",
        "  profileMatrix = np.zeros((4,len(motifs[0])))\n",
        "  # now just divide count matrix by number of motifs\n",
        "  for row in range(4):\n",
        "    for col in range(k):\n",
        "      profileMatrix[row][col] = pseudoMat[row][col] / len(motifs)\n",
        "  return profileMatrix\n",
        "\n",
        "def PseudoCountMatrix(countMat):\n",
        "  pseudo = [1*4]*len(countMat[0])\n",
        "  pseudoMatrix = pseudo+countMat\n",
        " #print(\"pseudo\", pseudo)\n",
        "  return pseudoMatrix\n",
        "\n",
        "\n",
        "def ProfileMostProbableKmer(genome, k, profile_matrix):\n",
        "  maxprob = 0\n",
        "  kmer = genome[0:k]\n",
        "  for i in range(1,len(genome) - k +1):\n",
        "    prob = 1\n",
        "    pattern = genome[i:i+k]\n",
        "    for j in range(k):\n",
        "      l = symbolToNumber(pattern[j])\n",
        "      prob *= profile_matrix[l][j]\n",
        "    if prob > maxprob:\n",
        "      maxprob = prob\n",
        "      kmer = pattern\n",
        "  return kmer\n",
        "  \n",
        "def symbolToNumber(symbol):\n",
        "\tif symbol == \"A\":\n",
        "\t\treturn 0\n",
        "\tif symbol == \"C\":\n",
        "\t\treturn 1\n",
        "\tif symbol == \"G\":\n",
        "\t\treturn 2\n",
        "\tif symbol == \"T\":\n",
        "\t\treturn 3\n",
        "\n",
        "def NumberToSymbol(number):\n",
        "\tif number == 0:\n",
        "\t\treturn \"A\"\n",
        "\tif number == 1:\n",
        "\t\treturn \"C\"\n",
        "\tif number == 2:\n",
        "\t\treturn \"G\"\n",
        "\tif number == 3:\n",
        "\t\treturn \"T\"\n",
        "\n",
        "def Consensus(profile):\n",
        "  consensus = \"\"\n",
        "  for col in range(len(profile[0])):  #column wise\n",
        "    max = 0\n",
        "    loc = 0\n",
        "    for row in range(4):\n",
        "      if profile[row][col] > max:\n",
        "        loc = row \n",
        "        max = profile[row][col]\n",
        "    consensus += NumberToSymbol(loc)  \n",
        "  return consensus\n",
        "\n",
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): #base case\n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem]\n",
        "    pref_pattern = numberToPattern(pref_index, k-1) \n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def Score(motifs):\n",
        "  profile = ProfileMatrix(motifs)\n",
        "  cons = Consensus(profile)\n",
        "  score = 0\n",
        "  for motif in motifs:  \n",
        "    for i in range(len(motif)): \n",
        "      if cons[i] != motif[i]:\n",
        "        score+=1\n",
        "  return score\n",
        "\n",
        "def GreedyMotifWithPseudoCount(genomes, k, t):\n",
        "  bestMotifs = BestMotifs(genomes, k,t)\n",
        "  kmers = generateTextKmer(genomes[0], k)\n",
        "  for kmer in kmers:\n",
        "    motifs = []\n",
        "    motifs.append(kmer)\n",
        "    for idx in range(1, t):\n",
        "      profile_matrix = ProfileMatrix(motifs) \n",
        "      motifs.append(ProfileMostProbableKmer(genomes[idx], k, profile_matrix))\n",
        "    \n",
        "    if Score(motifs) < Score(bestMotifs):\n",
        "      bestMotifs = motifs\n",
        "  return bestMotifs\n",
        "\n",
        "k = 3\n",
        "t = 5\n",
        "with open('/content/rosalind_ba2e.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "ans = GreedyMotifWithPseudoCount(genomelist, k, t)\n",
        "for i in ans:\n",
        "  print(i)"
      ],
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TTC\n",
            "ATC\n",
            "TTC\n",
            "ATC\n",
            "TTC\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AxOMHvvDR8jD"
      },
      "source": [
        "#### BA2F: Implement RandomizedMotifSearch"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MRT9m13wR8pt",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d10306c3-a275-4101-816a-3427b069be14"
      },
      "source": [
        "import numpy as np\n",
        "from random import randint\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "def CountMatrix(motifs): \n",
        "  count = np.zeros((4,len(motifs[0])))\n",
        "  for motif in motifs:\n",
        "    for i in range(len(motif)):\n",
        "      if motif[i-1] == 'A':\n",
        "        count[0][i-1] += 1\n",
        "      elif motif[i-1] == 'C':\n",
        "        count[1][i-1] += 1\n",
        "      elif motif[i-1] == 'G':\n",
        "        count[2][i-1] += 1\n",
        "      elif motif[i-1] == 'T':\n",
        "        count[3][i-1] += 1\n",
        "  return count\n",
        "\n",
        "def ProfileMatrix(motifs):\n",
        "  countMat = CountMatrix(motifs)\n",
        "  pseudoMat = PseudoCountMatrix(countMat)\n",
        "  profileMatrx = np.zeros((4,len(motifs[0])))\n",
        "  for row in range(4):\n",
        "    for col in range(k):\n",
        "      profileMatrx[row][col] = pseudoMat[row][col] / len(motifs)\n",
        "  return profileMatrx\n",
        "\n",
        "def PseudoCountMatrix(countMat):\n",
        "  pseudo = [1*4]*len(countMat[0])\n",
        "  pseudoMatrix = pseudo+countMat\n",
        "  return pseudoMatrix\n",
        "\n",
        "def ProfileMostProbableKmer(genome, k, profile_matrix):\n",
        "  maxprob = 0\n",
        "  kmer = genome[0:k]\n",
        "  for i in range(1,len(genome) - k +1):\n",
        "    prob = 1\n",
        "    pattern = genome[i:i+k]\n",
        "    for j in range(k):\n",
        "      l = symbolToNumber(pattern[j])\n",
        "      prob *= profile_matrix[l][j]\n",
        "    if prob > maxprob:\n",
        "      maxprob = prob\n",
        "      kmer = pattern\n",
        "  return kmer\n",
        "  \n",
        "def symbolToNumber(symbol):\n",
        "\tif symbol == \"A\":\n",
        "\t\treturn 0\n",
        "\tif symbol == \"C\":\n",
        "\t\treturn 1\n",
        "\tif symbol == \"G\":\n",
        "\t\treturn 2\n",
        "\tif symbol == \"T\":\n",
        "\t\treturn 3\n",
        "\n",
        "def NumberToSymbol(number):\n",
        "\tif number == 0:\n",
        "\t\treturn \"A\"\n",
        "\tif number == 1:\n",
        "\t\treturn \"C\"\n",
        "\tif number == 2:\n",
        "\t\treturn \"G\"\n",
        "\tif number == 3:\n",
        "\t\treturn \"T\"\n",
        "\n",
        "def Consensus(profile):\n",
        "  consensus = \"\"\n",
        "  for col in range(len(profile[0])):  \n",
        "    max = 0\n",
        "    loc = 0\n",
        "    for row in range(4):\n",
        "      if profile[row][col] > max:\n",
        "        loc = row  \n",
        "        max = profile[row][col]  \n",
        "    consensus += NumberToSymbol(loc)  \n",
        "  return consensus\n",
        "\n",
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): #base case\n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem] \n",
        "    pref_pattern = numberToPattern(pref_index, k-1) \n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def Score(motifs):\n",
        "  profile = ProfileMatrix(motifs)\n",
        "  cons = Consensus(profile)\n",
        "  score = 0\n",
        "  for motif in motifs: \n",
        "    for i in range(len(motif)): \n",
        "      if cons[i] != motif[i]:\n",
        "        score+=1\n",
        "  return score\n",
        "\n",
        "def getRandom(n):\n",
        "  return randint(0, n)\n",
        "\n",
        "def FormRandomlyMotifs(genomes, k):\n",
        "  motifs = []\n",
        "  for genome in genomes:\n",
        "    start = getRandom(len(genome)-k)\n",
        "    motifs.append(genome[start:start+k])\n",
        "  return motifs\n",
        "\n",
        "def generateMotifsSet(profile_matrix, k, genomes):\n",
        "  probKmers = []\n",
        "  for genome in genomes:\n",
        "    probKmers.append(ProfileMostProbableKmer(genome, k, profile_matrix))\n",
        "  return probKmers\n",
        "\n",
        "def RandomizedMotifSearch(genomes, k, t):\n",
        "  motifs = FormRandomlyMotifs(genomes, k)\n",
        "  bestMotifs = motifs\n",
        "  while True:\n",
        "    profile_matrix = ProfileMatrix(motifs)\n",
        "    motifs = generateMotifsSet(profile_matrix, k, genomes)\n",
        "    if Score(motifs) < Score(bestMotifs):\n",
        "      bestMotifs = motifs\n",
        "    else:\n",
        "      return bestMotifs\n",
        "\n",
        "\n",
        "k = 8\n",
        "t = 5\n",
        "with open('/content/rosalind_ba2f.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "  \n",
        "best_score = float(\"Inf\")\n",
        "best_result = []\n",
        "for i in range(1000):\n",
        "  ans = RandomizedMotifSearch(genomelist, k, t)\n",
        "  curr_score = Score(ans)\n",
        "  if curr_score <= best_score:\n",
        "    best_score = curr_score\n",
        "    best_result = ans\n",
        "for line in best_result:\n",
        "  print(line)"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TCTCGGGG\n",
            "CCAAGGTG\n",
            "TACAGGCG\n",
            "TTCAGGTG\n",
            "TCCACGTG\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "P5cl66_bR8wz"
      },
      "source": [
        "#### BA2G: Implement GibbsSampler"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "A7fQZP_KR83E",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7db5595a-09dc-4890-8072-524af955a5fa"
      },
      "source": [
        "import numpy as np\n",
        "import random\n",
        "from random import randint, choices\n",
        "\n",
        "def FormBestMotifs(genomes, k,t):\n",
        "  bmotifs = []\n",
        "  for genome in genomes:\n",
        "    bmotifs.append(genome[:k])\n",
        "  return bmotifs\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = []\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.append(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "\n",
        "def FormCountMatrix(motifs, k): \n",
        "  count = np.zeros((4,len(motifs[0])))\n",
        "  for motif in motifs:\n",
        "    for i in range(len(motif)):\n",
        "      if motif[i-1] == 'A':\n",
        "        count[0][i-1] += 1\n",
        "      elif motif[i-1] == 'C':\n",
        "        count[1][i-1] += 1\n",
        "      elif motif[i-1] == 'G':\n",
        "        count[2][i-1] += 1\n",
        "      elif motif[i-1] == 'T':\n",
        "        count[3][i-1] += 1\n",
        "  return count\n",
        "\n",
        "def FormProfile(motifs, k): \n",
        "  countMat = FormCountMatrix(motifs, k)\n",
        "  pseudoMat = PseudoCountMatrix(countMat, k)\n",
        "  profileMatrx = np.zeros((4,len(motifs[0])))\n",
        "  for row in range(4):\n",
        "    for col in range(k):\n",
        "      profileMatrx[row][col] = pseudoMat[row][col] / len(motifs)\n",
        "  return profileMatrx\n",
        "\n",
        "def PseudoCountMatrix(countMat, k): \n",
        "  pseudo = [1*4] * k   #np.ones((row, col)) here row = 4\n",
        "  pseudoMatrix = pseudo + countMat\n",
        "  return pseudoMatrix\n",
        "\n",
        "  \n",
        "def symbolToNumber(symbol):\n",
        "\tif symbol == \"A\":\n",
        "\t\treturn 0\n",
        "\tif symbol == \"C\":\n",
        "\t\treturn 1\n",
        "\tif symbol == \"G\":\n",
        "\t\treturn 2\n",
        "\tif symbol == \"T\":\n",
        "\t\treturn 3\n",
        "\n",
        "def NumberToSymbol(number):\n",
        "\tif number == 0:\n",
        "\t\treturn \"A\"\n",
        "\tif number == 1:\n",
        "\t\treturn \"C\"\n",
        "\tif number == 2:\n",
        "\t\treturn \"G\"\n",
        "\tif number == 3:\n",
        "\t\treturn \"T\"\n",
        "\n",
        "def Consensus(profile):\n",
        "  consensus = \"\"\n",
        "  for col in range(len(profile[0])):  \n",
        "    max = 0\n",
        "    loc = 0\n",
        "    for row in range(4):\n",
        "      if profile[row][col] > max:\n",
        "        loc = row  \n",
        "        max = profile[row][col]\n",
        "    consensus += NumberToSymbol(loc)  \n",
        "  return consensus\n",
        "\n",
        "\n",
        "def Score(motifs, k):\n",
        "  profile = FormProfile(motifs, k)\n",
        "  cons = Consensus(profile)\n",
        "  score = 0\n",
        "  for motif in motifs:  \n",
        "    for i in range(len(motif)): \n",
        "      if cons[i] != motif[i]:\n",
        "        score+=1\n",
        "  return score\n",
        "\n",
        "def getRandom(n):\n",
        "  return randint(0, n)\n",
        "\n",
        "def FormRandomlyMotifs(genomes, k):\n",
        "  motifs = []\n",
        "  for genome in genomes:\n",
        "    start = getRandom(len(genome)-k)\n",
        "    motifs.append(genome[start:start+k])\n",
        "  return motifs\n",
        "\n",
        "def generateMotifsSet(profile_matrix, k, genomes):\n",
        "  probKmers = []\n",
        "  for genome in genomes:\n",
        "    probKmers.append(ProfileMostProbableKmer(genome, k, profile_matrix))\n",
        "  return probKmers\n",
        "\n",
        "def probability(kmer, profile_matrix):\n",
        "  prob = 1\n",
        "  for j in range(len(kmer)):\n",
        "      l = symbolToNumber(kmer[j])\n",
        "      prob *= profile_matrix[l][j]\n",
        "  return prob\n",
        "\n",
        "def ProfileRandomGeneratedKmer(profile_matrix, k, genome):\n",
        "  prob = []\n",
        "  kmers = generateTextKmer(genome, k)\n",
        "  for kmer in kmers:\n",
        "    prob.append(probability(kmer, profile_matrix))\n",
        "  \n",
        "  motifs = random.choices(kmers, prob)\n",
        "  return motifs[0]\n",
        "\n",
        "\n",
        "def GibbsSampler(genomes, k, t, N):\n",
        "  motifs = FormRandomlyMotifs(genomes, k)\n",
        "  bestMotifs = motifs\n",
        "  for j in range(N):\n",
        "    i = getRandom(t-1)\n",
        "    profile_matrix = FormProfile(motifs[:i] + motifs[i+1 :], k)\n",
        "    motifi = ProfileRandomGeneratedKmer(profile_matrix, k, genomes[i])\n",
        "    motifs[i] = motifi \n",
        "    if Score(motifs, k) < Score(bestMotifs, k):\n",
        "      bestMotifs = motifs \n",
        "  return bestMotifs\n",
        "\n",
        "k = 8\n",
        "t = 5  \n",
        "N = 100 \n",
        "with open('/content/rosalind_ba2g.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "\n",
        "best = GibbsSampler(genomelist, k, t, N)\n",
        "best_score = Score(best, k)\n",
        "best_result = []\n",
        "for i in range(20):\n",
        "  ans = GibbsSampler(genomelist, k, t, N)\n",
        "  curr_score = Score(ans, k)\n",
        "  if curr_score < best_score:\n",
        "    best_score = curr_score\n",
        "    best_result = ans\n",
        "for line in best_result:\n",
        "    print(line)"
      ],
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "CCCTCTCG\n",
            "TAAGTGCC\n",
            "CCGAGACC\n",
            "CAGGTGCA\n",
            "CACGTGCA\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nYJizVXqR8-k"
      },
      "source": [
        "#### BA2H: Implement DistanceBetweenPatternAndStrings"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ca6eKtD8R9FL",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "55bd546b-d091-4e1b-b6f4-55040ad079b0"
      },
      "source": [
        "def DistanceBetweenPatternsandGenome(genomelist, pattern):\n",
        "  k = len(pattern)\n",
        "  distance = 0\n",
        "  for genome in genomelist:\n",
        "    hammingDistance = float(\"inf\")\n",
        "    genome_kmers = generateTextKmer(genome, k)\n",
        "    for kmer in  genome_kmers:\n",
        "      if hammingDistance > HammingDistance(pattern, kmer):\n",
        "        hammingDistance =  HammingDistance(pattern, kmer)\n",
        "    distance = distance+hammingDistance\n",
        "  #print(distance)  \n",
        "  return distance\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "def HammingDistance(text1, text2):\n",
        "    count=0;\n",
        "    for i in range(len(text1)):\n",
        "        if(text1[i] != text2[i]):\n",
        "            count+=1\n",
        "    return count\n",
        "\n",
        "\n",
        "#pattern = input()\n",
        "with open('/content/rosalind_ba2h.txt', 'r') as f:\n",
        "  pattern = f.readline().strip()\n",
        "  listWords = f.readline().strip().split(\" \")\n",
        "  #print(listWords)\n",
        "\n",
        "\n",
        "print(DistanceBetweenPatternsandGenome(listWords, pattern))\n"
      ],
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "5\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}